resolution/design/00_resolution.md

```diff
@@ -0,0 +1,331 @@
+[mfclass]: images/mfclass_diagram.png
+
+# Proposed Solution
+
+As part of the optimisation scheme a new class [`OptFunction`] will be created; objects of this class are not intended to handle instrumental resolution convolution.
+Instead resolution effects will be dealt with by extending `OptFunction` with one or more new classes.
+
+There has been some discussion of implementing new methods for estimating the instrumental resolution of a direct-geometry time-of-flight spectrometer, e.g., making use of Monte Carlo ray-tracing to produce a probability distribution of neutron states incident on the sample along with analytic approximations to the secondary spectrometer.
+
+Additionally, we have investigated new methods for performing the resolution convolution -- notably, reuse of in-resolution function evaluations for nearby detector pixels.
+
+In order to support multiple ways of estimating the instrument resolution and performing the convolution we require a flexible resolution convolution scheme to be implemented in conjunction with `OptFunction`.
+
+## `OptFunction` subclasses
+`OptFunction` will be extended to include an `initialise()` method which performs no action by default, and a `simulate()` method which calls the `initialise()` method before the `evaluate()` method. Same-named similar extensions will be added to `OptFunctions` as well.
+The two 'gateway' methods, `simulate()` and `optimise()`, in `OptModel` will be modified to first call its contained `OptFunctions` objects' `initialise()` and then to continue utilising a private `evaluate()` method to finish simulating or optimising the model.
+The `evaluate()` method of `OptFunction` and `OptFunctions` must remain public.
+
+With these changes we can then impelemnt a new subclass of `OptFunction` for any resolution convolution scheme.
+Notably, we must produce `TobyfitOptFunction` which reproduces the resolution estimation and convolution scheme in `tobyfit`.
+Other possible schemes include `RaytracedResolutionOptFunction` which uses a ray-tracing-derived lookup table for ![\delta\mathbf{k}_\text{i}];
+and `Kernel2DOptFunction` which uses a user-defined two dimensional kernel to 'smooth' an `axes` model in the same way that the independent data has been smoothed.
+
+![OptFunction_subclasses]
+
+[OptFunction_subclasses]: images/OptFunction_subclasses.png
+[`OptFunction`]: ../../optimisation/design/Model_Optimisation_Design.md#optimisation-function-object
+
+## Possible (partial) implementation
+The requisite methods for `OptModel`, `OptFunctions`, and `OptFunction` to support resolution covolution subclasses could be implemented as
+
+```matlab
+classdef OptModel
+    properties
+        multipliers {mustBeOptFunctions} = [];
+        foregrounds {mustBeOptFunctions} = [];
+        backgrounds {mustBeOptFunctions} = [];
+    end
+    methods
+        function varargout = simulate(obj, varargin)
+            obj.initialise();
+            varargout = obj.evaluate(varargin);
```

Comment:
Should it be `varargin{:}` in the `evaluate` calls?

Reply:
It's untested pseudo-code. But yes, I think you're right.

Reply:
"brain-compiled" is a phrase I read somewhere on Stack-overflow.